File: python/euler/bin/p0089.py
# project euler: problem 89

#   step 1:
#     IIIIIIIII     IX
#     XXXXXXXXX     XC
#     CCCCCCCCC     CM
#
#   step 2:
#     VIIII         IX
#     LXXXX         XC
#     DCCCC         CM
#
#   step 3:
#     IIIII         V
#     XXXXX         L
#     CCCCC         D
#
#   step 4:
#     IIII          IV
#     XXXX          XL
#     CCCC          CD

import re
from typing import TextIO


def replace_numbers(line: str) -> str:
    s = re.sub('IIIIIIIII|XXXXXXXXX|CCCCCCCCC', '##', line)  # step 1  IX/XC/CM
    s = re.sub('VIIII|LXXXX|DCCCC', '##', s)  # step 2  IX/XC/CM
    s = re.sub('IIIII|XXXXX|CCCCC', '#', s)  # step 3  V/L/D
    s = re.sub('IIII|XXXX|CCCC', '##', s)  # step 4  IV/XL/CD

    return s


def compute(fh: TextIO) -> str:
    acc = 0
    for line in fh.read().splitlines():
        acc += len(line) - len(replace_numbers(line))

    return str(acc)


def solve() -> str:
    from euler.lib.resource import asset_file

    fh = asset_file('p089_roman.txt')
    result = compute(fh)
    fh.close()
    return result

File: typescript/euler/bin/p0089.ts
// project euler: problem 89

/*
  step 1:
    IIIIIIIII     IX
    XXXXXXXXX     XC
    CCCCCCCCC     CM

  step 2:
    VIIII         IX
    LXXXX         XC
    DCCCC         CM

  step 3:
    IIIII         V
    XXXXX         L
    CCCCC         D

  step 4:
    IIII          IV
    XXXX          XL
    CCCC          CD

  We will need the following files to run this program.
    - https://projecteuler.net/project/resources/p089_roman.txt
*/

import { assetData } from "../lib/asset.ts";

function replaceNumbers(line: string): string {
  // step 1, 2, 3 and 4
  let s = line.replace(/IIIIIIIII|XXXXXXXXX|CCCCCCCCC/g, "##");
  s = s.replace(/VIIII|LXXXX|DCCCC/g, "##");
  s = s.replace(/IIIII|XXXXX|CCCCC/g, "#");
  s = s.replace(/IIII|XXXX|CCCC/g, "##");

  return s;
}

export const compute = (data: string): string => {
  function parseData(data: string): string[] {
    function splitLines(str: string): string[] {
      const result = str.split(/\r?\n/);
      if (result.at(-1) === "") {
        return result.slice(0, -1);
      } else {
        return result;
      }
    }

    return splitLines(data);
  }

  let acc = 0;
  for (const line of parseData(data)) {
    acc += line.length - replaceNumbers(line).length;
  }

  return String(acc);
};

export const solve = (): string => {
  const data = new TextDecoder().decode(assetData("p089_roman.txt"));
  return compute(data);
};
